### PR TITLE
ldso: Teach loadDynamicLibrary to work asynchronously under a flag

### PR DESCRIPTION
**Pull request by navytux**
In many places dynamic libraries are loaded synchronously, but for the
browser they are preloaded asynchronously (see 3446d2af "preload wasm
dynamic libraries when we can't load them synchronously"). However
preloading logic is placed outside loadDynamicLibrary.

We'll soon need to adjust dynamic linker logic and for a loaded library,
also first load its needed dynamic libraries. Thus, as the preparatory
step, move async loading logic into inside loadDynamicLibrary itself,
because otherwise we would need to teach the preloading code about
library dependencies too with much code duplication.

No need to introduce new test - asynchronous loading functionality is
verified by browser.test_dynamic_link with what was added there in 3446d2af.

### CODE REVIEW

**Comment by navytux** [comment]
( spawned from https://github.com/kripken/emscripten/pull/7512 )

**Comment by sbc100** [comment]
lgtm, but maybe @kripken should take a look too

File: src/support.js [hunk 1]
```diff
@@ -84,41 +84,102 @@ var asm2wasmImports = { // special asm2wasm imports
...
-function loadDynamicLibrary(lib) {
-  var libModule;
+// loadDynamicLibrary loads dynamic library @ lib URL / path.
+//
+// Several flags affect the loading:
+//
+// - if flags.loadAsync=true, the loading is performed asynchronously and
+//   loadDynamicLibrary returns corresponding promise.
+function loadDynamicLibrary(lib, flags) {
+  if (flags === undefined) {
```
Comment (kripken) [inline src/support.js]: can do `flags = flags || {}` which is a pattern we use elsewhere

Comment (navytux) [reply src/support.js]: sure, thanks for the advice.

**Comment by navytux** [comment]
@sbc100, @kripken, thanks for feedback. There is already a test that covers asynchronous loading functionality - it was added as part of 3446d2af (preload wasm dynamic libraries when we can't load them synchronously) to `browser.test_dynamic_link` here:

https://github.com/kripken/emscripten/commit/3446d2af#diff-b739639dfafbbd8097296133efb95a05R3019

The `preamble.js` is changed in the patch to use `loadDynamicLibrary(lib, {loadAsync: true})` instead of what was previously there and what is covered by abovementioned `browser.test_dynamic_link::"wasm in worker"` test. I think it should be already enough to have that test since here we only move-in-into-loadDynamicLibrary the functionality that was previously outside of it, but that is already tested.

Please let me know whether having that test is not enough and we should have one more for loadAsync.

For now I've amended the patch description with

```
No need to introduce new test - asynchronous loading functionality is
verified by browser.test_dynamic_link with what was added there in 3446d2af.
```

and adjusted the change itself with the following interdiff:

```diff
diff --git a/src/support.js b/src/support.js
index bfcb9af82..ed62ca3bf 100644
--- a/src/support.js
+++ b/src/support.js
@@ -103,10 +103,7 @@ function fetchBinary(url) {
 // - if flags.loadAsync=true, the loading is performed asynchronously and
 //   loadDynamicLibrary returns corresponding promise.
 function loadDynamicLibrary(lib, flags) {
-  if (flags === undefined) {
-    flags = {}
-  }
-
+  flags = flags || {};
 
   // libData <- lib
   function loadLibData() {
```

Hope it is ok.

Thanks,
Kirill

**Comment by kripken** [comment]
Those test failures look like something unrelated that is currently broken - can ignore it here.

I see what you mean about the testing, thanks - makes sense. Looks good!

**Comment by navytux** [comment]
@kripken, thanks for feedback and for applying the patch. @sbc100, thanks also for your review.

**Pull request merged**